aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -554,15 +816,131 @@ module aptos_framework::delegation_pool {
         account::get_signer_capability_address(&pool.stake_pool_signer_cap)
     }
 
+    /// Get the active share amount of the delegator.
+    fun get_delegator_active_shares(pool: &DelegationPool, delegator: address): u128 {
+        pool_u64::shares(&pool.active_shares, delegator)
+    }
+
+    /// Get the pending inactive share amount of the delegator.
+    fun get_delegator_pending_inactive_shares(pool: &DelegationPool, delegator: address): u128 {
+        pool_u64::shares(pending_inactive_shares_pool(pool), delegator)
+    }
+
+    /// Get the used voting power of a voter on a proposal.
+    fun get_used_voting_power(governance_records: &GovernanceRecords, voter: address, proposal_id: u64): u64 {
+        let votes = &governance_records.votes;
+        let key = VotingRecordKey {
+            voter,
+            proposal_id,
+        };
+        *smart_table::borrow_with_default(votes, key, &0)
+    }
+
+    /// Borrow the mutable used voting power of a voter on a proposal.
+    inline fun borrow_mut_used_voting_power(governance_records: &mut GovernanceRecords, voter: address, proposal_id: u64): &mut u64 {
+        let votes = &mut governance_records.votes;
+        let key = VotingRecordKey {
+            proposal_id,
+            voter,
+        };
+        smart_table::borrow_mut_with_default(votes, key, 0)
+    }
+
+    /// Update VoteDelegation of a delegator to up-to-date then borrow_mut it.
+    fun update_and_borrow_mut_delegator_vote_delegation(
+        pool: &DelegationPool,
+        governance_records :&mut GovernanceRecords,
+        delegator: address
+    ): &mut VoteDelegation {
+        let pool_address = get_pool_address(pool);
+        let locked_until_secs = stake::get_lockup_secs(pool_address);
+
+        let vote_delegation_table = &mut governance_records.vote_delegation;
+        // By default, a delegator's delegated voter is itself.
+        // TODO: recycle storage when VoteDelegation equals to default value.
```

Comment:
how do you recycle?

Reply:
When it equals to default value, remove the entry in the table. However doing it inside this function might not be ideal because this function always returns a `&mut`. Lots of storage need to be recycled in this module so I think it could be complicated and we will need a separate PR.